psi4/src/psi4/libmints/matrix.cc

```diff
@@ -3403,8 +3403,51 @@ SharedMatrix doublet(const SharedMatrix &A, const SharedMatrix &B, bool transA,
 
 SharedMatrix triplet(const SharedMatrix &A, const SharedMatrix &B, const SharedMatrix &C, bool transA, bool transB,
                      bool transC) {
-    SharedMatrix T = doublet(A, B, transA, transB);
-    SharedMatrix S = doublet(T, C, false, transC);
+
+    int nirrep = A->nirrep();
+    
+    if (nirrep != B->nirrep() || nirrep != C->nirrep()) {
+        throw PsiException("Input Matrices A, B, and C do not have the same number of irreps.", __FILE__, __LINE__);
+    }
+
+    // cost1 = cost of (AB)C
+    // cost2 = cost of A(BC)
+    int cost1 = 0;
+    int cost2 = 0;
+
+    for (int h = 0; h < nirrep; h++) {
+
+        int dim1 = transA ? A->colspi(h) : A->rowspi(h);
+        int dim2 = transA ? A->rowspi(h) : A->colspi(h);
+        int dim3 = transB ? B->colspi(h) : B->rowspi(h);
+        int dim4 = transB ? B->rowspi(h) : B->colspi(h);
+        int dim5 = transC ? C->colspi(h) : C->rowspi(h);
+        int dim6 = transC ? C->rowspi(h) : C->colspi(h);
```

Comment:
These checks still assume `A`, `B`, and `C` have the same symmetry ... determined from `XX->symmetry()`.  The underlying gemm code does know how to handle symmetry properly.  However, if it's too complicated to determine the cost analysis taking into account the different symmetries then I would wrap this check and the multiplications that follow in an if statement that checks for the symmetry of all the matrices.  Then In the case where the symmetries don't match I would default to the original triplet code and leave it to the caller to figure out.

What Lori had referenced [here](https://github.com/psi4/psi4/blob/87b0117c345f257ea763de92d0462e93e82e62d9/psi4/src/psi4/libmints/matrix.cc#L1359-L1378) is the style of what is needed to handle arbitrary symmetry.

Reply:
Andy, you may want to consider adding the extra symmetry check mentioned above and the extra tests mentioned below and wrapping up the PR there. It's still an improvement over the naive existing triplet, and per-irrep multiplication and reassembly could be revisited later.

Reply:
Got it!

Reply:
PR ready for review